src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -1222,17 +1221,37 @@ internal static void ReleaseExternalObjectsFromCurrentThread()
             IntPtr contextToken = GetContextToken();
 
             List<object> objects = new List<object>();
+
             foreach (GCHandle weakNativeObjectWrapperHandle in s_referenceTrackerNativeObjectWrapperCache)
             {
-                ReferenceTrackerNativeObjectWrapper? nativeObjectWrapper = Unsafe.As<ReferenceTrackerNativeObjectWrapper?>(weakNativeObjectWrapperHandle.Target);
-                if (nativeObjectWrapper != null &&
-                    nativeObjectWrapper._contextToken == contextToken)
+                // Here we aren't part of a GC callback, so other threads can still be running
+                // who are adding and removing from the collection. The collection we are enumerating over
+                // supports lock free enumeration, but it is possible we can run into handles
+                // that just got removed and freed. So we guard against that here. If they have
+                // been freed, we don't need to worry about them as they don't need to be released.
+                // In addition, we are only handling objects from our current thread here.
+                if (!weakNativeObjectWrapperHandle.IsAllocated)
+                {
+                    continue;
+                }
+
+                try
                 {
-                    objects.Add(nativeObjectWrapper._proxyHandle.Target);
+                    ReferenceTrackerNativeObjectWrapper? nativeObjectWrapper = Unsafe.As<ReferenceTrackerNativeObjectWrapper?>(weakNativeObjectWrapperHandle.Target);
+                    if (nativeObjectWrapper != null &&
+                        nativeObjectWrapper._contextToken == contextToken)
+                    {
+                        objects.Add(nativeObjectWrapper._proxyHandle.Target);
 
-                    // Separate the wrapper from the tracker runtime prior to
-                    // passing them.
-                    nativeObjectWrapper.DisconnectTracker();
+                        // Separate the wrapper from the tracker runtime prior to
+                        // passing them.
+                        nativeObjectWrapper.DisconnectTracker();
+                    }
+                }
+                catch (InvalidOperationException)
```

Comment:
This feels not great. Do we expect this to happen rarely enough that it's not worth worrying about? Admittedly, I can't really think of a way to avoid catching the exception here short of basically using `SafeHandle`-s in the map instead of `GCHandle`-s, or some other equivalent "ref counting with delayed dispose" thing. Maybe not worth it though.

Reply:
I believe this specific API is intended to be used only during shutdown / window close, so it wouldn't get called often and I believe it is unused in WinUI today (but they may use it for cleanup purposes in the future).  But to minimize how often this races, that is why I added the check earlier.

I can also do something like the below I think given we don't use pinning:

`GCHandle.InternalGet(GCHandle.ToIntPtr(weakNativeObjectWrapperHandle))`

Reply:
Replaced IsAllocated check and try catch to handle potential race with above approach.